src/core/UbqScaledImageGenerator.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2015 Google Inc.
+ *
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
+#include "include/core/UbqScaledImageGenerator.h"
+
+#include "src/core/SkPixmapPriv.h"
+
+auto UbqScaledImageGenerator::MakeFromEncoded(sk_sp<SkData> data, int maxImageSize)
+        -> std::unique_ptr<SkImageGenerator> {
+    auto codec = SkCodec::MakeFromData(data);
+    if (nullptr == codec) {
+        return nullptr;
+    }
+
+    return std::unique_ptr<SkImageGenerator>(
+            new UbqScaledImageGenerator(std::move(codec), data, maxImageSize));
+}
+
+static SkImageInfo adjust_info(SkCodec* codec, int maxImageSize) {
+    SkImageInfo info = codec->getInfo();
+    if (kUnpremul_SkAlphaType == info.alphaType()) {
+        info = info.makeAlphaType(kPremul_SkAlphaType);
+    }
+
+    if (SkEncodedOriginSwapsWidthHeight(codec->getOrigin())) {
+        info = SkPixmapPriv::SwapWidthHeight(info);
+    }
+
+    // Alter the advertised dimensions to fit into maxImageSize
+    if (info.width() > maxImageSize || info.height() > maxImageSize) {
+        float width = info.width();
+        float height = info.height();
+        float maxSize = maxImageSize;
+        const auto desiredScale = std::min(maxSize / width, maxSize / height);
+
+        // Query codec for supported dimensions
+        const auto supportedDimensions = codec->getScaledDimensions(desiredScale);
```

Comment:
How much smaller than the desired scale might this be? Do we need to add a lower limit here?

Reply:
Good point, I couldn't find a definitive answer. Could add a check that ensures that at least one edge has 0.75% of `maxSize`.

Reply:
Pushed a fix, that ensures at least 75% of `maxImageSize` is met.